src/librbd/image/RemoveRequest.cc

```diff
@@ -0,0 +1,910 @@
+  assert(!m_image_ctx->owner_lock.is_locked());
+
+  using klass = RemoveRequest<I>;
+  Context *ctx = create_context_callback<klass, &klass::handle_send_close_image>(this);
+
+  m_image_ctx->state->close(ctx);
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_send_close_image(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if (*result < 0) {
+    lderr(m_cct) << "error encountered while closing image: " << cpp_strerror(*result)
+                 << dendl;
+  }
+
+  switch_thread_context();
+  return nullptr;
+}
+
+template <typename I>
+void RemoveRequest<I>::switch_thread_context() {
+  ldout(m_cct, 20) << dendl;
+
+  using klass = RemoveRequest<I>;
+
+  Context *ctx = create_context_callback<klass, &klass::handle_switch_thread_context>(this);
+  m_op_work_queue->queue(ctx, 0);
+}
+
+template <typename I>
+void RemoveRequest<I>::handle_switch_thread_context(int r) {
+  ldout(m_cct, 20) << ": r=" << r << dendl;
+
+  delete m_image_ctx;
+  m_image_ctx = nullptr;
+
+  if (m_retval < 0) {
+    m_on_finish->complete(m_retval);
+    return;
+  }
+
+  remove_header();
+  return;
+}
+
+template<typename I>
+void RemoveRequest<I>::remove_header() {
+  ldout(m_cct, 20) << dendl;
+
+  using klass = RemoveRequest<I>;
+  librados::AioCompletion *rados_completion =
+    create_rados_ack_callback<klass, &klass::handle_remove_header>(this);
+  int r = m_ioctx.aio_remove(m_header_oid, rados_completion);
+  assert(r == 0);
+  rados_completion->release();
+}
+
+template<typename I>
+Context *RemoveRequest<I>::handle_remove_header(int *result) {
+  ldout(m_cct, 20) << ": r=" << *result << dendl;
+
+  if ((*result < 0) && (*result != -ENOENT)) {
+    lderr(m_cct) << "error removing header: " << cpp_strerror(*result) << dendl;
+    return m_on_finish;
+  }
+
+  remove_image();
+  return nullptr;
+}
+
+template<typename I>
+void RemoveRequest<I>::remove_image() {
+  ldout(m_cct, 20) << dendl;
+
+  if (m_old_format || m_unknown_format) {
+    remove_v1_image();
+  } else {
+    remove_v2_image();
+  }
+}
+
+template<typename I>
+void RemoveRequest<I>::remove_v1_image() {
+  ldout(m_cct, 20) << dendl;
+
+  Context *ctx = new FunctionContext([this] (int r) {
+      r = tmap_rm(m_ioctx, m_image_name);
+      handle_remove_v1_image(r);
+    });
+
+  m_op_work_queue->queue(ctx, 0);
+}
+
+template<typename I>
+void RemoveRequest<I>::handle_remove_v1_image(int r) {
+  ldout(m_cct, 20) << ": r=" << r << dendl;
+
+  m_old_format = (r == 0);
+  if ((r < 0) && !m_unknown_format) {
+    if (r != -ENOENT) {
+      lderr(m_cct) << "error removing image from v1 directory: " << cpp_strerror(r) << dendl;
+    }
+
+    m_on_finish->complete(r);
+    return;
+  }
+
+  if (!m_old_format) {
+    remove_v2_image();
+    return;
+  }
+
+  m_on_finish->complete(0);
+}
+
+template<typename I>
+void RemoveRequest<I>::remove_v2_image() {
+  ldout(m_cct, 20) << dendl;
+
+  remove_journal();
+}
+
+template<typename I>
+void RemoveRequest<I>::remove_journal() {
```

Comment:
If you disable journaling and object map at the start -- we can also eliminate the duplicate code for cleanup here and just let the update features state machine do it all for us.